### PR TITLE
Sign shreds on gpu without a stage.

### PR DESCRIPTION
**Pull request by aeyakovenko**
#### Problem

1. Signing takes CPU time
2. Shred are allocated using the regular Vec heap

#### Summary of Changes

1. Allocate shreds directly into Packets which use a PinnedVec/Recycler
2. Sign packets on the GPU

Todo:
- [x] RAM usage is 40%? (@sagar-solana Resident ram seems fine)
- [x] broadcast seems to be much slower with CPU sigverify (not sure if its just warmup)

Separate PR
- [ ] separate the entries_to_shreds as a stage
- [ ] avoid converting back to a Shred inside blocktree insert

Fixes #

### CODE REVIEW

**Comment by aeyakovenko** [comment]
@sagar-solana @sakridge @pgarg66 
i think its ready for review

File: ledger/src/shred.rs [hunk 1]
```diff
@@ -425,67 +552,136 @@ impl Shredder {
...
+            .iter()
+            .map(|s| (*s, self.keypair.pubkey().to_bytes()))
+            .collect();
+        pubkeys.insert(std::u64::MAX, [0u8; 32]);
+        let mut privkeys: HashMap<u64, [u8; 32]> = slots
+            .iter()
+            .map(|s| (*s, self.keypair.secret.to_bytes()))
+            .collect();
+        privkeys.insert(std::u64::MAX, [0u8; 32]);
+        //sigverify_shreds::sign_shreds_gpu(&mut data_shreds, &pubkeys, &privkeys, recycler_cache);
```
Comment (aeyakovenko) [inline ledger/src/shred.rs]: @sakridge I can't figure out why this doesn't work. "test_multi_fec_block_coding" fails, and so does the network partition test.

**Comment by aeyakovenko** [comment]
@sakridge recycler ends up with 38k entries in the gc for some reason

**Comment by aeyakovenko** [comment]
@sagar-solana are you sure the RAM stats that we see on the dashboard are resident memory and not virtual?   When I ssh into the validators I see about 10-15% resident memory used.

**Comment by sagar-solana** [comment]
@aeyakovenko the script is pretty simple. It's just using top to subtract used memory from total memory. I think that probably shows disk cache that Linux sets up. I should probably switch to using free so that it's more accurate. It's definitely reporting more than resident memory used by processes right now. I'll update it to use free when I get a chance.

File: ledger/src/shred.rs [hunk 2]
```diff
@@ -427,98 +554,165 @@ impl Shredder {
...
+    ) -> (Vec<Packets>, Vec<Packets>, u32) {
+        let (mut data_shreds, last_shred_index) = self.entries_to_unsigned_data_shreds(
+            recycler_cache,
+            entries,
+            is_last_in_slot,
+            next_shred_index,
+        );
+
+        let now = Instant::now();
+        let slots = Self::read_slots(&data_shreds);
```
Comment (pgarg66) [inline ledger/src/shred.rs]: All shreds generated here will have the same slot. We do not need to read it from the shreds. We can use `self.slot`

File: ledger/src/shred.rs [hunk 3]
```diff
@@ -427,98 +554,165 @@ impl Shredder {
...
+        let mut pubkeys: HashMap<u64, [u8; 32]> = slots
+            .iter()
+            .map(|s| (*s, self.keypair.pubkey().to_bytes()))
+            .collect();
+        pubkeys.insert(std::u64::MAX, [0u8; 32]);
```
Comment (pgarg66) [inline ledger/src/shred.rs]: Why is this needed?

File: ledger/src/shred.rs [hunk 4]
```diff
@@ -427,98 +554,165 @@ impl Shredder {
...
+                    .par_iter()
+                    .filter_map(|shred_data_batch| {
                         Shredder::generate_coding_shreds(
+                            recycler_cache,
                             self.slot,
                             self.fec_rate,
                             shred_data_batch,
                             self.version,
                         )
+                        //let mut packets = Packets::new_with_recycler(
```
Comment (pgarg66) [inline ledger/src/shred.rs]: Delete this?

File: ledger/src/shred.rs [hunk 5]
```diff
@@ -592,44 +806,27 @@ impl Shredder {
...
-                    Shred {
-                        common_header,
-                        data_header: DataShredHeader::default(),
-                        coding_header,
-                        payload,
-                    }
-                })
-                .collect()
+        }
+        if fec_rate != 0.0 {
```
Comment (pgarg66) [inline ledger/src/shred.rs]: if rate is `0.0`, we should not even generate the coding shreds.

File: ledger/src/shred.rs (hunk 3 above)
Comment (aeyakovenko) [reply ledger/src/shred.rs]: The memory that we send to the GPU is flat, so we don't to reorder or skip any of the packets.  if there is an error in the data it points to MAX/0..., which should fail the signature.
